File: src/core/crypto/hmac_handler.py
import hmac
import hashlib
import secrets
from typing import Tuple
from uuid import UUID

from src.config import get_settings


class HMACHandler:
    def __init__(self, secret_key: str):
        self.secret_key = secret_key.encode("utf-8")

    def generate_signature(self, certificate_id: UUID, nonce: str) -> str:
        """Generate HMAC signature for QR URL"""
        payload = f"{certificate_id}|{nonce}"
        signature = hmac.new(
            self.secret_key, payload.encode("utf-8"), hashlib.sha256
        ).hexdigest()
        return signature

    def verify_signature(
        self, certificate_id: UUID, nonce: str, signature: str
    ) -> bool:
        """Verify HMAC signature"""
        expected = self.generate_signature(certificate_id, nonce)
        return hmac.compare_digest(expected, signature)

    def generate_nonce(self) -> str:
        """Generate cryptographically secure nonce for QR"""
        return secrets.token_urlsafe(32)


def get_hmac_handler() -> HMACHandler:
    return HMACHandler(get_settings().secret_key)

File: app/Services/Crypto/HMACHandler.php
<?php

namespace App\Services\Crypto;

class HMACHandler
{
    protected string $secretKey;

    /**
     * Pass your application's private signing key via the constructor.
     */
    public function __construct(string $secretKey)
    {
        $this->secretKey = $secretKey;
    }

    /**
     * Generate HMAC signature for QR URL.
     * Matches Python: hmac.new(key, payload, hashlib.sha256).hexdigest()
     */
    public function generateSignature(string $certificateId, string $nonce): string
    {
        $payload = "{$certificateId}|{$nonce}";

        return hash_hmac('sha256', $payload, $this->secretKey);
    }

    /**
     * Verify HMAC signature securely using a constant-time comparison algorithm.
     * Matches Python: hmac.compare_digest()
     */
    public function verifySignature(string $certificateId, string $nonce, string $signature): bool
    {
        $expected = $this->generateSignature($certificateId, $nonce);

        return hash_equals($expected, $signature);
    }

    /**
     * Generate a cryptographically secure, URL-safe nonce for QR codes.
     * Matches Python: secrets.token_urlsafe(32)
     */
    public function generateNonce(): string
    {
        // 32 bytes of raw randomness equals roughly 43 characters when Base64URL encoded
        $bytes = random_bytes(32);
        
        // Convert to URL-safe Base64 (strip padding and swap safe URL characters)
        return rtrim(strtr(base64_encode($bytes), '+/', '-_'), '=');
    }
}